core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java

```diff
@@ -1023,6 +1054,16 @@ VirtualHost build(VirtualHostBuilder template) {
         assert extensions != null;
         assert blockingTaskExecutor != null;
 
+        final EventLoopGroup serviceWorkerGroup;
+        final boolean shutdownServiceWorkerGroupOnStop;
+        if (this.serviceWorkerGroup != null) {
+            serviceWorkerGroup = this.serviceWorkerGroup;
+            shutdownServiceWorkerGroupOnStop = this.shutdownServiceWorkerGroupOnStop;
+        } else {
+            serviceWorkerGroup = template.serviceWorkerGroup;
+            shutdownServiceWorkerGroupOnStop = template.shutdownServiceWorkerGroupOnStop;
```

Comment:
`template.serviceWorkerGroup` can be `null` when a user never called `ServerBuilder.serviceWorkerGroup()`. If that's the case, we should use `ServerBuilder.workerGroup` as a fallback, so we preserve the old behavior(`workerGroup == serviceWorkerGroup`):

```suggestion
            if (template.serviceWorkerGroup != null) {
                serviceWorkerGroup = template.serviceWorkerGroup;
                shutdownServiceWorkerGroupOnStop = template.shutdownServiceWorkerGroupOnStop;
            } else {
                serviceWorkerGroup = serverBuilder.workerGroup;
            }
```

By doing this, we're sure `VirtualHost.serviceWorkerGroup` is never null once fully constructed.

Note: You'll have to make `ServerBuilder.workerGroup` package-private.